### PR TITLE
handle vnode with undefined data

### PR DESCRIPTION
**Pull request by yarom82**
There was a run time error when used with vnode with undefined data.
API allows to provide undefined data.

### CODE REVIEW

File: test/eventlisteners.js [hunk 1]
```diff
@@ -169,4 +169,13 @@ describe('event listeners', function() {
     elm.firstChild.click();
     assert.equal(3, result.length);
   });
+  it('handle vnode with undefined data', function () {
```
Comment (yarom82) [inline test/eventlisteners.js]: Could be redundant as we have this test at test/core.js

File: test/style.js [hunk 2]
```diff
@@ -116,6 +116,15 @@ describe('style', function() {
     fakeRaf.step();
     assert.equal(elm.style.fontSize, '20px');
   });
+  it('handle vnode with undefined data', function() {
```
Comment (yarom82) [inline test/style.js]: Could be redundant as we have this test at test/core.js

**Comment by caridy** [comment]
Can you explain the rationale about this? Why will you want to have a vnode without data (even if it is empty)? Snabbdom is built for speed, `h()` is just an artifact for people to build an object that represent a vnode, but you can build your own `h` with your own logic. For example, I normally have my own `h()` that returns the same empty object (`Object.create(null)`) for data.

**Comment by mightyiam** [comment]
@caridy either some variant of this PR or make [`data`](https://github.com/snabbdom/snabbdom/blob/7b45afea49a32e2f4b0c5f12ca7df0040fd1dcf7/src/vnode.ts#L15) always `VNodeData`.

**Comment by caridy** [comment]
@mightyiam I'm reluctant to merge this PR because it includes many checks that must be performed during the critical path, when the same thing can be achieve in user-land by adding a simple abstraction around `h()` to produce valid vnodes. I don't think snabbdom should be in the business of validation, which is something that I have mentioned many times before.

**Comment by mightyiam** [comment]
Then would it be fine to type the `data` property of `VNode` as always `VNodeData` instead of `VNodeData | undefined`? Otherwise, what seems to be a valid `VNode`, with `data: undefined`, causes a runtime error.

**Comment by caridy** [comment]
@mightyiam yes, that seems fine. we can probably adjust `h()` to enforce that data cannot be `undefined`, to eliminate all kind of critical path checks.

**Comment by staltz** [comment]
I agree with @caridy. Let's not merge it (or not merge it *yet*).

**Comment by yarom82** [comment]
Closing in favor of PR #329

**Pull request closed without merging**